tools/check.py

```diff
@@ -87,343 +80,421 @@ def _validate_no_fixme(self):
+        if (not re.match(r"^((https?|ftp)://)", address, re.IGNORECASE)
+                and not re.match(r"^mailto:", address)):
+            dest = address.split("#")
+            if len(dest) > 1:
+                anchor = dest[1]
+            else:
+                anchor = None
+            dest = dest[0] or self.filename
+            dest_path = os.path.join(self.lesson_dir, dest)
 
-        return check_text, no_check_text
+            # If HTML file need to check for Markdown file.
+            if re.search(r"\.(html?)$", dest_path, re.IGNORECASE):
+                dest_path = dest_path.replace(".html", ".md")
+
+                if not os.path.isfile(dest_path):
+                    logging.error(
+                        "In {0}: "
+                        "The document links to {1}, but could not find "
+                        "the expected Markdown file {2}".format(
+                            self.filename, address, dest_path))
+                    return False
+                elif anchor:
+                    if not anchor in MarkdownValidator(dest_path).ast.anchors:
+                        logging.error(
+                            "In {0}: "
+                            "The document links to {1}, but could not find "
+                            "the anchor {2} in {3}".format(
+                                self.filename, address, anchor, dest_path))
+                        return False
+            else:
+                if not os.path.isfile(dest_path):
+                    logging.error(
+                        "In {0}: "
+                        "Could not find the linked asset file "
+                        "{1}. If this is a URL, it must be "
+                        "prefixed with http(s):// or ftp://.".format(
+                            self.filename, dest_path))
+                    return False
+
+        return True
 
     def _validate_links(self):
-        """Validate all references to external content
+        """Validate links.
 
-        This includes links AND images: these are the two types of node that
-        CommonMark assigns a .destination property"""
-        check_text, no_check_text = self._partition_links()
+        Verify that all the links in the document are valid."""
+        res = True
+        links = []  # List of ('text', 'address')
 
-        valid = True
-        for link_node in check_text:
-            res = self._validate_one_link(link_node, check_text=True)
-            valid = valid and res
+        def get_links(key, val, format, meta):
+            if key in ("Link", "Image"):
+                links.append(
+                    (vh.ast_to_string(val[0]),
+                     val[1][0]))
 
-        for link_node in no_check_text:
-            res = self._validate_one_link(link_node, check_text=False)
-            valid = valid and res
-        return valid
+        pandocfilters.walk(self.ast.body, get_links, "", {})
```

Comment:
How does this function decide whether to check the link text against the document title?

Last time I checked, some templates allowed free-form links (eg `[click here](destination.md)`). The old validator had a separate function that figured out which links needed their text validated, but I see the function was removed.

Has the rule allowing free-form links been changed?